docs/make.md

```diff
@@ -0,0 +1,186 @@
+
+Finally, you'll need to be familiar with the command line interface (CLI).  If you haven't seen it yet, take a look at the [Command Line Tutorial](cli.html).
+
+
+## Example program
+We'll work in a directory called `make-example` just for this tutorial.
+```console
+$ mkdir make-example
+```
+
+Create a file called `main.cpp` and copy-paste this Hello World program.
+```c++
+#include <iostream>
+using namespace std;
+
+int main() {
+  cout << "Hello World!\n";
+}
+```
+{: data-title="main.cpp" }
+
+Create a file called `Makefile` and copy-paste this code.
+```make
+CXX ?= g++
+CXXFLAGS ?= -Wall -Werror -pedantic -g --std=c++11 -Wno-sign-compare -Wno-comment
+
+# Run regression test
+test: main.exe
+	./main.exe < main_test.in > main_test.out
+	diff main_test.out main_test.out.correct
+	echo PASS
+
+# Compile the main executable
+main.exe: main.cpp
+	$(CXX) $(CXXFLAGS) main.cpp -o main.exe
+
+# Remove automatically generated files
+clean :
+	rm -rvf *.exe *~ *.out *.dSYM *.stackdump
+```
+{: data-title="Makefile" }
+
+Your directory should look like this:
+```console
+$ tree
+.
+├── Makefile
+└── main.cpp
+```
+
+
+## Compile
+Compile manually, then remove the executable.
+```console
+$ g++ -Wall -Werror -pedantic -g --std=c++11 -Wno-sign-compare -Wno-comment main.cpp -o main.exe
+$ ./main.exe
+Hello World!
+$ rm main.exe
+```
+
+Compile using `make`.  Notice that it automated the long compile command.
+```console
+$ make main.exe
+g++ -Wall -Werror -pedantic -g --std=c++11 -Wno-sign-compare -Wno-comment main.cpp -o main.exe
+```
+
+Run the executable created by the previous `make` command.
+```console
+$ ./main.exe
+Hello World!
+```
+
+Clean up using `make`.  Notice that it automated the `rm` command.
+```console
+$ make clean
+rm -rvf *.exe *~ *.out *.dSYM *.stackdump
+removed 'main.exe'
+```
+
+
+## Test
+Makefiles are useful for automating tests, sometimes called a *regression test*.
+
+Put input (stdin) in a file.  We'll use `main_test.in` in this example.
+```
+this program doesn't read stdin,
+so this text doesn't matter
+```
+{: data-title="main_test.in" }
+
+Put the correct output in a file.  We'll use `main_test.out.correct`.  In EECS 280 and EECS 281, a few correct output files are typically provided.
+```
+Hello World!
+```
+{: data-title="main_test.out.correct" }
+
+Run manually with [input redirection](cli.html#input-redirection-) and [output redirection](cli.html#output-redirection-).  Compare with [`diff`](cli.html#diff).  No output means the files match.
+```console
+$ ./main.exe < main_test.in > main_test.out
+$ diff main_test.out main_test.out.correct
+```
+
+Run the regression test automatically with `make`.
+```console
+$ make test
+./main.exe < main_test.in > main_test.out
+diff main_test.out main_test.out.correct
+```
+
+<div class="primer-spec-callout info" markdown="1">
+**Pro-tip:** Run commands in parallel with `make -j`.
```

Comment:
I didn't know this!

Reply:
It makes the tests on P2 and P5 in particular run MUCH faster!